### PR TITLE
feat: property-based scheduling: updated the IMC controller to support property provider integration

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR adds support for property provider in the IMC controller.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### Special notes for your reviewer

- A behavior change: the IMC controller now reports the "agent is healthy" condition by probing the K8s API server (originally healthy == can list nodes)
- Some tests are kept out to control PR size; will submit them in future PRs.

### CODE REVIEW

**Comment by zhiying-lin** [comment]
left comments on the implementation and have not reviewed the test yet

File: cmd/memberagent/main.go
Comment (michaelawyu) [reply cmd/memberagent/main.go:348]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:52]: We have to use the global context as the Start() method for the property provider is called with a context that will exit when the member agent exits to allow graceful shutdown; if we pass the context from Reconcile() instead, the property provider will exit as soon as the first reconciliation call ends.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:43]: Done.

File: apis/cluster/v1beta1/membercluster_types.go
Comment (michaelawyu) [reply apis/cluster/v1beta1/membercluster_types.go:151]: At this moment there is no plan to support multiple property providers running at the same time, even though technically speaking it's obviously doable. I think it might be a little bit weird to use the `AllClusterPropertyProvidersStarted` reference now, but it could make sense if we decide to support so.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:73]: At this moment there is no plan to support multiple property providers running at the same time (the planned use case is one per platform).

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:291]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:41]: The config is for the property provider. Comment added.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:71]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:384]: I am not 100% sure if I understand the question correctly, but the workqueue has guaranteed that for any individual key enqueued it will not be processed concurrently; in addition, there is only one InternalMemberCluster object visible to the member agent any way.

And the behavior is as described: it will either wait until a response is returned in time or time out (and fall back to the default mechanism). The processing has already started before the `select` statement runs.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:73]: Discussed offline.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (zhiying-lin) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go:445]:
```suggestion
			if err := r.reportClusterPropertiesWithPropertyProvider(ctx, tc.imc); err == nil { // if NO error
				t.Fatalf("reportClusterPropertiesWithPropertyProvider(), got no error, want error")
			}
```
nit

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (zhiying-lin) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go:392]: redundant field?

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (zhiying-lin) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go:468]: redundant field?

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (zhiying-lin) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go:637]:
```suggestion
			if err := r.reportClusterPropertiesWithPropertyProvider(ctx, tc.imc); err != nil {
				t.Fatalf("reportClusterPropertiesWithPropertyProvider(), got error %v, want no error", err)
```

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:382]: The point was not to terminate the go routine -> there's no way in Go that we could achieve that, but to limit the max. # of go routines we will create for running the Collect() call.

Comment (zhiying-lin) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:382]: the current comment is confusing. maybe correct the comment to mention you want to limit the max # of go routines?

We have already configured the deadline (1s) and it also helps.

I don't have strong opinion on adding this check though as it does not hurt.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go]: Done.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller_test.go]: Done.

**Review by zhiying-lin — APPROVED** [review]

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (ryanzhang-oss) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller.go:322]: does the sync.Once retry?

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (ryanzhang-oss) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller.go:316]: This is basically spawn a goroutine and wait for it to finish inline. I don't understand why not just call the function inline with the childCtx?

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (ryanzhang-oss) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller.go:414]: why does collect not return an error?

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (ryanzhang-oss) [inline pkg/controllers/internalmembercluster/v1beta1/member_controller.go:413]: same question as the above, why use async when there is only one goroutine? the main thread is blocking anyway. What's the benefit?

File: cmd/memberagent/main.go
Comment (michaelawyu) [reply cmd/memberagent/main.go:342]: Yeah, it will be sent in a separate PR as the provider PR hasn't been merged yet.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:277]: Yeah, this is probing the API server at the health endpoint. For other components it might be better to explore later as at this moment the health condition we add is largely unused. The pressure should be light.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:322]: Nah, at this moment it's try once only. I could add retry if you'd like

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:316]: This is to add a deadline to the Start() method, which is calling external implementation with no guarantees, i.e., it might block and not respect the cancellation of `childCtx`, hence the select call.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:414]: Collect() is not defined to return an error in the interface; the reason is that the response already features conditions where detailed info. regarding collection failures can be logged and exposed.

File: pkg/controllers/internalmembercluster/v1beta1/member_controller.go
Comment (michaelawyu) [reply pkg/controllers/internalmembercluster/v1beta1/member_controller.go:413]: Yeah, similarly, Collect() is implemented externally, and on the caller side there's no guarantee that the call would not block and get cancelled promptly if it timed out.

**Comment by michaelawyu** [comment]
Hi Ryan! I will merge this PR first to unblock the progress; I will submit another PRs to address any concerns/changes if you find them necessary

**Pull request merged**
